wasm/leb128/read.go

```diff
@@ -7,66 +7,81 @@
 package leb128
 
 import (
+	"errors"
 	"io"
 )
 
-// ReadVarUint32 reads a LEB128 encoded unsigned 32-bit integer from r, and
-// returns the integer value, and the error (if any).
-func ReadVarUint32(r io.Reader) (uint32, error) {
-	var (
-		b     = make([]byte, 1)
-		shift uint
-		res   uint32
-		err   error
-	)
-	for {
-		if _, err = io.ReadFull(r, b); err != nil {
-			return res, err
+// read uN defined in https://webassembly.github.io/spec/core/binary/values.html#binary-int
+// note: the result type is uint64, so N must <= 64.
+func readVarUint(r io.Reader, N uint) (uint64, error) {
+	p := make([]byte, 1)
+	_, err := r.Read(p)
+	if err != nil {
+		return 0, err
+	}
+	n := uint64(p[0])
+	if n < 1<<7 && n < 1<<N {
+		return n, nil
+	} else if n >= 1<<7 && N > 7 {
+		m, err := readVarUint(r, N-7)
```

Comment:
the original implementation was using a loop.
now we are using recursion (at which Go isn't particularly good at), plus we allocate a 1-len slice everytime.

couldn't we switch back to a loop?
or at least pass a "work" slice as argument to `readVarUint`?

Reply:
I have implemented a loop version.
the specification format has some differences with https://en.wikipedia.org/wiki/LEB128.
allowing trialing zeros, but having upper bound encoding size makes it hard to get right.
this recursion version is pretty close to the formula definition of the specification. so I left it as a baseline to test the loop one.